### ISSUE TITLE
depth_to_pointcloud.py focal length and principal point in pixels or something else?

### ISSUE THREAD

**Issue by karantai** [LiheYoung/Depth-Anything#127]
I have used the `dense_to_pointcloud.py` script in order to generate the point clouds of images, but they are highly distorted.

![Screenshot from 2024-03-15 18-54-55](https://github.com/LiheYoung/Depth-Anything/assets/82314287/9bb7e7b9-a5f0-4457-8c61-3f8cec451b19)

Do I have to convert my camera intrinsics, that is focal length and principal point to other measurement units or is it something else?

The code is provided by @1ssb:

```
# infer.py
# Code by @1ssb
import argparse
import os
import glob
import torch
import numpy as np
from PIL import Image
import torchvision.transforms as transforms
import open3d as o3d
from tqdm import tqdm
from zoedepth.models.builder import build_model
from zoedepth.utils.config import get_config

# Global settings
FL = 715.0873
FY = 23691.920027359894
FX = 23705.79001167295
NYU_DATA = False
FINAL_HEIGHT = 6336
FINAL_WIDTH = 9504
P_x, P_y = 4820.05051308297, 3127.5650602109763
INPUT_DIR = '~/pictures_2023_09_21-15_51_24_1'
OUTPUT_DIR = '~/Depth-Anything/my_test/output'
DATASET = 'nyu' # Lets not pick a fight with the model's dataloader


def process_images(model):
    if not os.path.exists(OUTPUT_DIR):
        os.makedirs(OUTPUT_DIR)

    image_paths = glob.glob(os.path.join(INPUT_DIR, '*.png')) + glob.glob(os.path.join(INPUT_DIR, '*.JPG'))
    for image_path in tqdm(image_paths, desc="Processing Images"):
        try:
            color_image = Image.open(image_path).convert('RGB')
            original_width, original_height = color_image.size
            image_tensor = transforms.ToTensor()(color_image).unsqueeze(0).to('cuda' if torch.cuda.is_available() else 'cpu')

            pred = model(image_tensor, dataset=DATASET)
            if isinstance(pred, dict):
                pred = pred.get('metric_depth', pred.get('out'))
            elif isinstance(pred, (list, tuple)):
                pred = pred[-1]
            pred = pred.squeeze().detach().cpu().numpy()

            # Resize color image and depth to final size
            resized_color_image = color_image.resize((FINAL_WIDTH, FINAL_HEIGHT), Image.LANCZOS)
            resized_pred = Image.fromarray(pred).resize((FINAL_WIDTH, FINAL_HEIGHT), Image.NEAREST)

            focal_length_x, focal_length_y = (FX, FY) if not NYU_DATA else (FL, FL)
            x, y = np.meshgrid(np.arange(FINAL_WIDTH), np.arange(FINAL_HEIGHT))
            x = (x - P_x) / focal_length_x
            y = (y - P_y) / focal_length_y
            z = np.array(resized_pred)
            points = np.stack((np.multiply(x, z), np.multiply(y, z), z), axis=-1).reshape(-1, 3)
            colors = np.array(resized_color_image).reshape(-1, 3) / 255.0

            pcd = o3d.geometry.PointCloud()
            pcd.points = o3d.utility.Vector3dVector(points)
            pcd.colors = o3d.utility.Vector3dVector(colors)
            o3d.io.write_point_cloud(os.path.join(OUTPUT_DIR, os.path.splitext(os.path.basename(image_path))[0] + ".pcd"), pcd)
        except Exception as e:
            print(f"Error processing {image_path}: {e}")

def main(model_name, pretrained_resource):
    config = get_config(model_name, "eval", DATASET)
    config.pretrained_resource = pretrained_resource
    model = build_model(config).to('cuda' if torch.cuda.is_available() else 'cpu')
    model.eval()
    process_images(model)

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument("-m", "--model", type=str, default='zoedepth', help="Name of the model to test")
    parser.add_argument("-p", "--pretrained_resource", type=str, default='local::./checkpoints/depth_anything_metric_depth_indoor.pt', help="Pretrained resource to use for fetching weights.")

    args = parser.parse_args()
    main(args.model, args.pretrained_resource)
```

**Comment by 1ssb** [comment]
YES, oh my god how is your focal length that large?

**Comment by karantai** [comment]
What do you mean by yes? Now they are in pixels units!

**Comment by 1ssb** [comment]
I am not sure, I have never seen such large focal lengths. Try denominations of 10 to figure out what is meaningful to you. If your photo is indeed idk 12000 odd units long, I do not know what to advise you. I have no experience with such images.

**Comment by karantai** [comment]
Thank you for your advise.
But in general, using you your code, the measurement units should be in pixels or something else? And the results of the point cloud will be in the corresponding measurement units?

**Comment by 1ssb** [comment]
The focal lengths are all metric, i do not know how else to answer your question.

**Comment by karantai** [comment]
Thank you very much

**Comment by karantai** [comment]
I convert all the pixel units to mm based on the the pixel size of my sensor, and also I used both outdoor and indoor models to check the results. Also, I converted the output the of the model (z) to 1/z and now the results even though they are wrong make more sense, as expected. Also, my focal length is 85 mm.

Yet, in the point cloud below the only thing that make sense are the pipes on the left of the point cloud. Is there any explanation for the distorted point cloud? Maybe the model is not trained for the specific rgb images that I use?

![Screenshot from 2024-03-19 15-15-53](https://github.com/LiheYoung/Depth-Anything/assets/82314287/289d8b02-47b0-4bbf-84b6-dab92542823a)

**Comment by 1ssb** [comment]
The model outputs depth not disparity. I suggest going through the model code once again and understanding the focal lengths and the image size representations.

**Comment by karantai** [comment]
My results were correct from the beginning, but my viewpoint was wrong. I believe that the only way to correctly visualize the point cloud is to view it from the camera perspective, that is the viewpoint of the camera when it captured the scene, as shown below. 

In order to transform the point cloud generated by a single image to global coordinates, one should apply shifts (Xo, Yo, Zo) and rotations regarding the start of the global coordinate system, right? And then the point cloud will be 'global', that is it will be viewed as if it was captured by a lidar?
  
![Screenshot from 2024-03-27 13-04-13](https://github.com/LiheYoung/Depth-Anything/assets/82314287/ef9405fd-373c-427c-bd49-417b3721a26c)

**Comment by 1ssb** [comment]
Were you not using a 3D application?

**Comment by karantai** [comment]
No I am not. I am using single RGB images, I use depth_to_pointcloud.py to derive the depth map and construct a point cloud, which is from my understanding is relative to the camera. Now I am trying to derive the global point cloud using the exterior orientation of the camera in that moment. Probably, I am missing something.

Thank you for your immediate responses

**Comment by 1ssb** [comment]
To visualize a point cloud it is imperative that you use a 3D application like Meshlab and correct scales. So no you are wrong, even though your visualization looks more correct to you now the fx an fy values along with your width and depth need to be divided by a denomination of 10. 

The Meshlab allows you to manipulate the rotation and translation interactively. So no writing scripts for visualisation.
